R/types-chr.R

```diff
@@ -0,0 +1,295 @@
+#' # all the time. Here is an example with the names attribute:
+#' latin1 <- string(c(0x63, 0x61, 0x66, 0xE9), "latin1")
+#' latin1 <- set_names(latin1)
+#'
+#' # The names attribute is encoded in latin1 as we would expect:
+#' str_encoding(names(latin1))
+#'
+#' # However the names are converted to UTF-8 by the c() function:
+#' str_encoding(names(c(latin1)))
+#' str_bytes(names(c(latin1)))
+#'
+#' # Bad things happen when the encoding marker is lost and R performs
+#' # a conversion. R will assume that the string is encoded according
+#' # to the current locale:
+#' \dontrun{
+#' bad <- set_names(str_set_encoding(latin1, "unknown"))
+#' set_utf8_locale()
+#'
+#' str_encoding(names(c(bad)))
+#' str_bytes(names(c(bad)))
+#' }
+chr_set_encoding <- function(x, encoding = c("unknown", "UTF-8", "latin1", "bytes")) {
+  Encoding(x) <- match.arg(encoding)
+  x
+}
+#' @rdname chr_set_encoding
+#' @export
+chr_encoding <- function(x) {
+  Encoding(x)
+}
+#' @rdname chr_set_encoding
+#' @export
+str_set_encoding <- function(x, encoding = c("unknown", "UTF-8", "latin1", "bytes")) {
+  stopifnot(is_string(x))
+  chr_set_encoding(x, encoding)
+}
+#' @rdname chr_set_encoding
+#' @export
+str_encoding <- function(x) {
+  stopifnot(is_string(x))
+  Encoding(x)
+}
+
+#' Set the locale's codeset for testing.
+#'
+#' Setting a locale's codeset (specifically, the \code{LC_CTYPE}
+#' category) produces side effects in R's handling of strings. The
+#' most important of these affects how the R parser marks strings. R
+#' has specific internal support for latin1 (single-byte encoding) and
+#' UTF-8 (multi-bytes variable-width encoding) strings. If the locale
+#' codeset is latin1 or UTF-8, the parser will mark all strings with
+#' the corresponding encoding. It is important for strings to have
+#' consistent encoding markers, as they determine a number of side
+#' effects in the handling of strings (see
+#' \code{\link{str_set_encoding}()}).
+#'
+#' UTF-8 is the only variable-width encoding supported in R.
+#' Multi-byte fixed-width encodings like UCS-2 are supported by R, but
+#' supplying UTF-16 strings (the default Unicode encoding on Windows)
+#' will cause the parser to choke on the character delimiters (though
+#' you may not see that behaviour in Windows IDEs if they translate
+#' inputs to another encoding for you). Note finally that
+#' \code{set_utf8_locale()} will not work on Windows as only latin1
+#' and MBCS locales are supported on this OS.
+#'
+#' Note that these helpers are only provided for testing interactively
+#' the effects of changing locale codeset. They let you quickly change
+#' the default text encoding to latin1, UTF-8, or non-UTF-8 MBCS. They
+#' are not widely tested and do not provide a way of setting the
+#' language and region of the locale. They have permanent side effects
+#' and should not be used in package examples, unit tests, or in the
+#' course of a data analysis.
+#'
+#' @export
```

Comment:
when I do it on my computer, it's returning the new locale, not the previous. But in the R CMD check scripts in `tools::`, they do use it as if it returned the old locale. Maybe a bug introduced at some point without anyone noticing?